java-filmorate-sprint11-try-1

constantineiljinyh opened the pull request.

constantineiljinyh committed:
java-filmorate-sprint11-try-1

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -15,19 +16,19 @@
-    @NotBlank(message = "Электронная почта не может быть пустой")
-    @Email(message = "Неправильный формат электронной почты")
-    private String email;
-
```
Ниже написать не могу
Коллекция с друзьями пользователей более не требуется, так как хранение мы осуществляем в БД)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -1,28 +1,37 @@
+    private final LikeFilmsStorage likeFilms;
+
+    @Autowired
+    public FilmServiceImpl(Storage<Film> filmStorage, UserService userService, LikeFilmsStorage likeFilms) {
```
Может заменить на аннотацию?

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -1,28 +1,37 @@
 import java.util.List;
+import java.util.stream.Collectors;
 
+@Transactional
```
Что делает данная аннотация? Корректно ли ты ее применил на весь сервис?
В других сервисах тоже самое

TarasovSam replied:
Ты не ответил на вопрос (ну это ладно)
Главное, что не внес необходимые правки (не все методы вносят изменения в БД)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -48,49 +57,49 @@ public Film getFilmById(Integer filmId) {
-            log.error("Ошибка обновления фильма: {}", e.getMessage());
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage());
+        if (!filmStorage.isExist(updatedFilm.getId())) {
+            throw new NotFoundException("Фильм не найден.");
```
Логирование?
И далее по коду

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -48,49 +57,49 @@ public Film getFilmById(Integer filmId) {
+                .limit(count)
+                .collect(Collectors.toList());
     }
 
```
В строке 114 предварительно нужно проверить на null

TarasovSam replied:
Не внес правки

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/RatingMPAServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+
+@Transactional
+@Service
+public class RatingMPAServiceImpl implements ClassificationService<RatingMPA> {
```
Логирование

TarasovSam replied:
Не внес логирование в сервис

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/RatingMPAServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final ClassificationStorage<RatingMPA> ratingMpa;
+
+    public RatingMPAServiceImpl(RatingMPAStorageDbImpl ratingMPADbStorage) {
```
Аннотация

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -3,24 +3,28 @@
+    private final FriendsStorage friendsStorage;
 
-    public InMemoryUserService(Storage<User> userStorage) {
+    public UserServiceImpl(Storage<User> userStorage, FriendsStorage friendsStorage) {
```
Аннотация?
И далее по коду

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -74,66 +78,50 @@ public void addFriend(Integer userId, Integer friendId) {
         return userStorage.remove(userId);
     }
 
+    @Override
```
Здесь прописал аннотацию переопределения, а для остальных методов нет)

TarasovSam replied:
Это тоже не внес

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/UserServiceImpl.java:
```diff
@@ -41,15 +45,15 @@ public List<User> getAllUsers() {
+        return updated;
     }
 
     public User getUserById(Integer userId) {
```
А если userId не существует?

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorageDbImpl.java:
```diff
@@ -0,0 +1,181 @@
+
+@Component
+@Primary
+public class FilmStorageDbImpl implements Storage<Film> {
```
В принципе уже можно перейти в названии от Storage к Repository
Реализацию класса Storage тоже можно убрать (разделить две репозитория так сказать)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -1,10 +1,11 @@
 import ru.yandex.practicum.filmorate.exception.NotFoundException;
 import ru.yandex.practicum.filmorate.model.Film;
+import ru.yandex.practicum.filmorate.storage.Storage;
 
```
Ниже написать не могу
Эта реализация более не будет использовать, ее по сути можно исключить)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/friend/FriendsStorage.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public interface FriendsStorage {
```
Интерфейс для репозитория можно не выделять)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -1,11 +1,12 @@
 import ru.yandex.practicum.filmorate.exception.ValidationException;
 import ru.yandex.practicum.filmorate.model.User;
+import ru.yandex.practicum.filmorate.storage.Storage;
 
```
Ниже написать не могу 
Эта реализация более не будет использовать, ее по сути можно исключить)

TarasovSam commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,49 @@
+    id_ratingMPA INTEGER not null references ratingMPA (id_ratingMPA) on delete cascade on update cascade
+);
+
+create TABLE if not exists users (
```
Можно было бы ввести ограничение на уникальность логина и почты на уровне таблицы в БД)

create unique index if not exists USER_EMAIL_UINDEX on USERS (email);
create unique index if not exists USER_LOGIN_UINDEX on USERS (login);

TarasovSam commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,4 @@
+
+INSERT INTO genre(name_genre) VALUES ('Комедия'), ('Драма'), ('Мультфильм'), ('Триллер'), ('Документальный'), ('Боевик');
```
Заполнение таблицы без id?

TarasovSam commented on src/test/java/ru/yandex/practicum/filmorate/storage/film/FilmStorageDbImplTest.java:
```diff
@@ -0,0 +1,263 @@
+
+    @Test
+    public void testAddFilm() {
+        Film filmToAdd = new Film();
```
Создание объектов лучше перенести в блок BeforeEach
Т.е. создание всех объектов будет в одном блоке, а в тестирование будет только логика тестирования

TarasovSam reviewed (commented):
Константин, привет!
Проверил твое задание, все довольно неплохо. Оставил комментарии с тем, что можно улучшить и что нужно поправить, плюс пару вопросов)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

constantineiljinyh committed:
java-filmorate-sprint11-try-2
Теперь жанры хранятся в Set<>,
Заменил все конструкторы аннотации,
Убрал коллекцию с друзьями
Изменил создание объектов в тестах, где это удобно.

constantineiljinyh commented:
Семен привет! Заполнение таблицы без id? Не стал делать потому что база данных генерирует. Исправил. Почему так не стоит делать?

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -41,7 +41,8 @@ public class Film {
     private Integer rate;
 
-    private List<Genre> genres;
+    private Set<Genre> genres;
```
Используй LinkedHashSet вместо простого Set

TarasovSam reviewed (commented):
Константин, проверил твое задание. Ты внес не все изменения, оставил соответствующие 
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

constantineiljinyh committed:
java-filmorate-sprint11-try-3
Убрал интерфейсы для репозитория
Теперь жанры хранятся в LinkedHashSet<>,
Поставил аннотации Transactional только в методы которые вносят изменения в БД,
Добавил логирование и доп проверки.

constantineiljinyh committed:
java-filmorate-sprint11-try-3
Убрал интерфейсы для репозитория
Теперь жанры хранятся в LinkedHashSet<>,
Поставил аннотации Transactional только в методы которые вносят изменения в БД,
Добавил логирование и доп проверки.

constantineiljinyh committed:
java-filmorate-sprint11-try-3
Убрал интерфейсы для репозитория
Теперь жанры хранятся в LinkedHashSet<>,
Поставил аннотации Transactional только в методы которые вносят изменения в БД,
Добавил логирование и доп проверки.

constantineiljinyh commented:
Привет, сорян. Я забыл, что там еще внутри комментарии могут быть). Transactional просто в начале так поставил, решил не влияет не на что). Над методами же ставят, что бы не путать которые вносят изменения, а которые нет. А то, что он весит над всеми ничего не ломает, как и Override. Я ни к тому что мне пофигу, а просто не предал значения и забыл)

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -96,7 +112,13 @@ public List<Film> getPopularFilms(int count) {
+        }
         log.info("Удаление фильма с ID {} ", filmId);
         return filmStorage.remove(filmId);
     }
```
Теперь уже в строке 130 предварительно нужно проверить на null

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -9,25 +9,26 @@
 import java.util.List;
 import java.util.stream.Collectors;
 
-@Transactional
```
Да, когда ставишь аннотацию над классом, то она относится ко всем методам, что в данном случае не корректно
@Transactional нужно обозначать только те методы, которые вносят изменения в БД, а для тех, которые не вносят, можно прописать @Transactional(readonly=true), чтобы указать ее очевидно (как и с @Override)

TarasovSam reviewed (commented):
Константин, проверил твое задание, спасибо за готовность дорабатывать код. Оставил еще пару комментариев))
Задание я принимаю, успеха в обучении и до следующего спринта!) Если будут вопросы, пиши в пачку)
